woocommerce/woocommerce-gutenberg-products-block
# Store API: Add shipping rate api

Shipping Rates API which allows you to retrieve shipping quotes for a given address for the current cart.

Closes #1318

Address is passed to the request, and at least country is required in order to return rates. There may be multiple arrays of rates if someone is using the 'packages' and splitting the cart into chunks. Items in the package are returned in the response you they can be referenced to the cart itself.

Example request:
```
http://local.wordpress.test/wp-json/wc/store/cart/shipping-rates?country=US&state=AL
```

Example response:

```json
[
  {
    "destination": {
      "address_1": null,
      "address_2": null,
      "city": null,
      "state": "AL",
      "postcode": null,
      "country": "US"
    },
    "items": [
      "6512bd43d9caa6e02c990b0a82652dca",
      "c20ad4d76fe97759aa27a0c99bff6710"
    ],
    "shipping-rates": [
      {
        "name": "International",
        "description": "Posted via UPS international.",
        "delivery_time": "Delivered in 7-14 days",
        "price": "20.00",
        "rate_id": "flat_rate:3",
        "instance_id": 3,
        "method_id": "flat_rate",
        "meta_data": [
          {
            "key": "Items",
            "value": "Beanie &times; 1, Belt &times; 1"
          }
        ]
      }
    ]
  }
]
```

For a cart with multiple packages, it would be split like this:

```json
[
  {
    "destination": {
      "address_1": null,
      "address_2": null,
      "city": null,
      "state": "AL",
      "postcode": null,
      "country": "US"
    },
    "items": [
      "6512bd43d9caa6e02c990b0a82652dca"
    ],
    "shipping-rates": [
      {
        "name": "International",
        "description": "Posted via UPS international.",
        "delivery_time": "Delivered in 7-14 days",
        "price": "20.00",
        "rate_id": "flat_rate:3",
        "instance_id": 3,
        "method_id": "flat_rate",
        "meta_data": [
          {
            "key": "Items",
            "value": "Beanie &times; 1"
          }
        ]
      }
    ]
  },
  {
    "destination": {
      "address_1": null,
      "address_2": null,
      "city": null,
      "state": "AL",
      "postcode": null,
      "country": "US"
    },
    "items": [
      "c20ad4d76fe97759aa27a0c99bff6710"
    ],
    "shipping-rates": [
      {
        "name": "International",
        "description": "Posted via UPS international.",
        "delivery_time": "Delivered in 7-14 days",
        "price": "20.00",
        "rate_id": "flat_rate:3",
        "instance_id": 3,
        "method_id": "flat_rate",
        "meta_data": [
          {
            "key": "Items",
            "value": "Belt &times; 1"
          }
        ]
      }
    ]
  }
]
```

`delivery_time` and `description` are not fields that are currently included in Woo Core. In my test branch I have introduced them and I will be raising a core PR for feedback in due time so we can support that.  

## Testing this PR

- Ensure you have configured Woo shipping zones and have some shipping methods
- Add some items to your cart. You can confirm shipping is working on your cart page.
- Now do an API request using basic auth with the same user. You can customise the address fields if you wish: `http://local.wordpress.test/wp-json/wc/store/cart/shipping-rates?country=US&state=AL`
- Check the response is accurate.
